crates/core/src/types/block.rs

```diff
@@ -56,13 +61,56 @@ impl RLPEncode for BlockHeader {
 
 // The body of a block on the chain
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Body {
+pub struct BlockBody {
     transactions: Vec<Transaction>,
+    // TODO: ommers list is always empty, so we can remove it
     ommers: Vec<BlockHeader>,
     withdrawals: Vec<Withdrawal>,
 }
 
-impl RLPEncode for Body {
+impl BlockBody {
+    pub const fn empty() -> Self {
+        Self {
+            transactions: Vec::new(),
+            ommers: Vec::new(),
+            withdrawals: Vec::new(),
+        }
+    }
+
+    pub fn compute_transactions_root(&self) -> H256 {
+        let transactions_iter: Vec<_> = self
+            .transactions
+            .iter()
+            .enumerate()
+            .map(|(i, tx)| {
+                // TODO: check if tree is RLP encoding the value
+
+                // Key: RLP(tx_index)
+                let mut k = Vec::new();
+                i.encode(&mut k);
+
+                // Value: tx_type || RLP(tx)  if tx_type != 0
+                //                   RLP(tx)  else
+                let mut v = Vec::new();
+                match tx {
+                    // Legacy transactions don't have a prefix
+                    Transaction::LegacyTransaction(_) => {}
+                    _ => v.push(tx.tx_type()),
+                }
+
+                tx.encode(&mut v);
```

Comment:
shouldn't `tx.encode` return the encoded `(type, payload)` so that it abstracts the type?

Reply:
Good idea! Since this encoding isn't RLP, having it like that would be really confusing for users. However, a `Transaction::encode_with_type` method would be useful.